snap-update-tools/check_snap_availability.py

```diff
@@ -0,0 +1,218 @@
+    response = requests.get(url, headers=headers)
+    if response.status_code != 200:
+        raise SystemExit(
+            f"Failed to get info about {snap_spec.name} from the snap store."
+        )
+
+    return response.json()
+
+
+def is_snap_available(snap_spec: SnapSpec, store_response: dict) -> bool:
+    """
+    Process the response from the snap store and check whether the specified
+    snap is available.
+    :param snap_spec: the snap specification
+    :param store_response: the response from the snap store
+    :return: True if the snap is available, False otherwise
+    """
+    # let's split the "channel" provided by the user
+    # into the track and the risk
+
+    def matches_spec(channel_info: dict) -> bool:
+        # some of the channel names don't follow the "track/risk" format
+        # so we need to handle them separately
+        if "/" not in snap_spec.channel:
+            return (
+                channel_info["channel"]["name"] == snap_spec.channel
+                and channel_info["version"] == snap_spec.version
+                and channel_info["channel"]["architecture"] in snap_spec.arch
+            )
+        else:
+            track, risk = snap_spec.channel.split("/")
+            return (
+                channel_info["channel"]["track"] == track
+                and channel_info["channel"]["risk"] == risk
+                and channel_info["version"] == snap_spec.version
+                and channel_info["channel"]["architecture"] in snap_spec.arch
+            )
+
+    return any(
+        matches_spec(channel_info)
+        for channel_info in store_response["channel-map"]
+    )
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Check whether snaps are available in the snap store."
+    )
+    parser.add_argument("version", help="Version of the snaps to check for.")
+    parser.add_argument(
+        "yaml_file",
+        type=argparse.FileType("r"),
+        help="Path to the YAML file specifying the snap requirements.",
+    )
+    parser.add_argument(
+        "--timeout",
+        help="Timeout in seconds after which the program will stop checking.",
+        default=300,
+        type=float,
+    )
+    args = parser.parse_args()
+
+    yaml_content = yaml.load(args.yaml_file, Loader=yaml.FullLoader)
+
+    # create the matrix of all combinations of the specified characteristics
+    snap_specs = [
+        SnapSpec(snap["name"], args.version, channel, arch)
+        for snap in yaml_content["required-snaps"]
+        for channel in snap["channels"]
+        for arch in snap["architectures"]
+    ]
+
+    timeout = 60.0
+    # Dict to store whether each snap is available.
+    already_available = {snap_spec: False for snap_spec in snap_specs}
+
+    # Set the deadline.
+    deadline = time.time() + timeout
+
+    while True:
+        # Record of snaps for which we've already fetched the data from the store.
+        already_queried = set()
```

Comment:
What is the purpose of `already_queried`? Are there inherently duplicated items in `snap_specs`? if there are, should we remove them above? something like `snap_specs = list(set(snap_specs))`. I think I'm missing something here

Reply:
that's an artifact from the previous version, trimmed  in https://github.com/canonical/hwcert-jenkins-tools/commit/f5493d726e52fabdf733474266ef20a51fb931f9